### PR TITLE
[ML] Simplify reverting to snapshot while datafeed is reassigning

### LINKED ISSUE #65710: MlDistributedFailureIT testClusterWithTwoMlNodes_RunsDatafeed_GivenOriginalNodeGoesDown failure
**Build scan**:
* https://gradle-enterprise.elastic.co/s/owwgiqfcfvmpm
* https://gradle-enterprise.elastic.co/s/szmkmg6lirwbs
* https://gradle-enterprise.elastic.co/s/el775yclzmwtu

**Repro line**:
> ./gradlew ':x-pack:plugin:ml:internalClusterTest' --tests "org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT.testClusterWithTwoMlNodes_RunsDatafeed_GivenOriginalNodeGoesDown" -Dtests.seed=EDD33680F671DE6F -Dtests.security.manager=true -Dtests.locale=en-US -Dtests.timezone=Europe/Paris -Druntime.java=8

**Reproduces locally?**:
No.

**Applicable branches**:
7.x. 
It fails at master branch as well, but with a different error message. I am not sure whehther they are related.
The build scan for master branch failure is: https://gradle-enterprise.elastic.co/s/rcoohbqmucb4e
and error message is:
```
java.lang.AssertionError: |  
-- | --
  | Expected: a value equal to or greater than <80000L> |  
  | but: <69165L> was less than <80000L> |  

at org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT.testClusterWithTwoMlNodes_RunsDatafeed_GivenOriginalNodeGoesDown(MlDistributedFailureIT.java:487)
--
```

**Failure history**:

The error is an IOException happened during teardown time. Normally, I would consider this type of error transient and let it pass. However this is a new test that just added and it [started to fail](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-30d,mode:quick,to:now))&_a=(columns:!(_source),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:b646ed00-7efc-11e8-bf69-63c8ef516157,key:build.branch,negate:!t,params:!(pull-request-2,pull-request-1),type:phrases,value:'pull-request-2,%20pull-request-1'),query:(bool:(minimum_should_match:1,should:!((match_phrase:(build.branch:pull-request-2)),(match_phrase:(build.branch:pull-request-1))))))),index:b646ed00-7efc-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:testClusterWithTwoMlNodes_RunsDatafeed_GivenOriginalNodeGoesDown),sort:!(process.time-start,desc))) at the same day. So I am creating this issue in case there are some underlying issue about it.

**Failure excerpt**:
```

java.io.IOException: Node didn't close within 10 seconds. |  
-- | --


at __randomizedtesting.SeedInfo.seed([EDD33680F671DE6F:4A9E02922FD46A84]:0) |  
-- | --
  |   | at org.elasticsearch.test.InternalTestCluster$NodeAndClient.close(InternalTestCluster.java:1091) |  
  |   | at org.elasticsearch.test.InternalTestCluster.stopNodesAndClients(InternalTestCluster.java:1854) |  
  |   | at org.elasticsearch.test.InternalTestCluster.stopNodesAndClient(InternalTestCluster.java:1844) |  
  |   | at org.elasticsearch.test.InternalTestCluster.stopNode(InternalTestCluster.java:1728) |  
  |   | at org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT.testClusterWithTwoMlNodes_RunsDatafeed_GivenOriginalNodeGoesDown(MlDistributedFailureIT.java:497) |  
  |   | at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) |  
  |   | at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) |  
  |   | at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) |  
  |   | at java.lang.reflect.Method.invoke(Method.java:498) |  
  |   | at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750) |  
  |   | at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938) |  
  |   | at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974) |  
  |   | at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988) |  
  |   | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  |   | at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49) |  
  |   | at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45) |  
  |   | at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48) |  
  |   | at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64) |  
  |   | at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47) |  
  |   | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  |   | at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368) |  
  |   | at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817) |  
  |   | at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468) |  
  |   | at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947) |  
  |   | at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832) |  
  |   | at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883) |  
  |   | at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894) |  
  |   | at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45) |  
  |   | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  |   | at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41) |  
  |   | at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40) |  
  |   | at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40) |  
  |   | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  |   | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  |   | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  |   | at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53) |  
  |   | at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47) |  
  |   | at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64) |  
  |   | at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54) |  
  |   | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  |   | at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368) |  
  |   | at java.lang.Thread.run(Thread.java:748)



```

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
Previously, we were trying to compare that the job had gone past
its current model snapshot in order to determine whether we
should perform recovery. However, this is not necessary. We can
instead simply always perform recovery if the opening job
has a running datafeed task. This greatly simplifies the code.

Relates #65630

Fixes #65710

### CODE REVIEW

**Review by benwtrent — APPROVED** [review]

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/task/OpenJobPersistentTasksExecutor.java [hunk 1]
```diff
@@ -200,16 +201,15 @@ private void runJob(JobTask jobTask, JobState jobState, OpenJobAction.JobParams
...
+                    // This job has a running datafeed attached to it.
                     // In order to prevent gaps in the model we revert to the current snapshot deleting intervening results.
-                    ModelSnapshot modelSnapshot = datafeedContext.getModelSnapshot() == null ?
-                        ModelSnapshot.emptySnapshot(jobTask.getJobId()) : datafeedContext.getModelSnapshot();
-                    logger.info("[{}] job had advanced past its current model snapshot [{}]; performing recovery",
-                        jobTask.getJobId(), modelSnapshot.getSnapshotId());
-                    revertToSnapshot(modelSnapshot, ActionListener.wrap(
+                    String modelSnapshotId = params.getJob().getModelSnapshotId() == null ?
+                        ModelSnapshot.EMPTY_SNAPSHOT_ID : params.getJob().getModelSnapshotId();
+                    logger.info("[{}] job has running datafeed task; performing recovery", jobTask.getJobId());
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/task/OpenJobPersistentTasksExecutor.java:211]: Can you change "performing recovery" to something like "reverting to last good model snapshot [X]"

But only if `modelSnapshotId != ModelSnapshot.EMPTY_SNAPSHOT_ID`

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/task/OpenJobPersistentTasksExecutor.java:211]: Right after that log message, we'll get the logging from the revert action which follows up and explains the recovery. I thought that was nice. Does it make sense or you still prefer to mention we're going to revert in there?

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/task/OpenJobPersistentTasksExecutor.java (hunk 1 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/task/OpenJobPersistentTasksExecutor.java]: OK, changed it to "reverting to current snapshot"

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

**Review by davidkyle — APPROVED** [review]
LGTM

Nice job deleting code

**Pull request merged**